v3_async_wip/tests/test_iothub_session.py

```diff
@@ -0,0 +1,1289 @@
+
+    @pytest.mark.it(
+        "Sets the provided `device_id` and `module_id` values on the IoTHubClientConfig used to create the IoTHubMQTTClient"
+    )
+    @pytest.mark.parametrize("shared_access_key, sastoken_fn, ssl_context", create_auth_params)
+    @pytest.mark.parametrize("device_id, module_id", create_id_params)
+    async def test_ids(
+        self, mocker, device_id, module_id, shared_access_key, sastoken_fn, ssl_context
+    ):
+        spy_mqtt_cls = mocker.spy(mqtt, "IoTHubMQTTClient")
+
+        IoTHubSession(
+            hostname=FAKE_HOSTNAME,
+            device_id=device_id,
+            module_id=module_id,
+            shared_access_key=shared_access_key,
+            sastoken_fn=sastoken_fn,
+            ssl_context=ssl_context,
+        )
+
+        cfg = spy_mqtt_cls.call_args[0][0]
+        assert cfg.device_id == device_id
+        assert cfg.module_id == module_id
+
+    @pytest.mark.it(
+        "Sets the provided `ssl_context` on the IoTHubClientConfig used to create the IoTHubMQTTClient, if provided"
+    )
+    @pytest.mark.parametrize(
+        "shared_access_key, sastoken_fn, ssl_context", create_auth_params_custom_ssl
+    )
+    @pytest.mark.parametrize("device_id, module_id", create_id_params)
+    async def test_custom_ssl_context(
+        self, mocker, device_id, module_id, shared_access_key, sastoken_fn, ssl_context
+    ):
+        spy_mqtt_cls = mocker.spy(mqtt, "IoTHubMQTTClient")
+
+        IoTHubSession(
+            hostname=FAKE_HOSTNAME,
+            device_id=device_id,
+            module_id=module_id,
+            shared_access_key=shared_access_key,
+            sastoken_fn=sastoken_fn,
+            ssl_context=ssl_context,
+        )
+
+        cfg = spy_mqtt_cls.call_args[0][0]
+        assert cfg.ssl_context is ssl_context
+
+    @pytest.mark.it(
+        "Sets a default SSLContext on the IoTHubClientConfig used to create the IoTHubMQTTClient, if `ssl_context` is not provided"
+    )
+    @pytest.mark.parametrize(
+        "shared_access_key, sastoken_fn, ssl_context", create_auth_params_default_ssl
+    )
+    @pytest.mark.parametrize("device_id, module_id", create_id_params)
+    async def test_default_ssl_context(
+        self, mocker, device_id, module_id, shared_access_key, sastoken_fn, ssl_context
+    ):
+        assert ssl_context is None
+        spy_mqtt_cls = mocker.spy(mqtt, "IoTHubMQTTClient")
+        my_ssl_context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
+        original_ssl_ctx_cls = ssl.SSLContext
+
+        # NOTE: due to implementation of SSLContext, we have to mock, and then unset the mock.
+        # Thus, this side effect has been implemented to do just that.
+        def return_and_reset(*args, **kwargs):
+            ssl.SSLContext = original_ssl_ctx_cls
+            assert kwargs["protocol"] is ssl.PROTOCOL_TLS_CLIENT
+            return my_ssl_context
+
+        mocker.patch.object(ssl, "SSLContext", side_effect=return_and_reset)
```

Comment:
This feels clunky. I've been staring at this for a while and I still have no idea what you're doing here. The comment `due to the implementation of SSLContext` doesn't tell me anything and I think it adds to my confusion.

Reply:
Yeah, I remember thinking that as I wrote it.... I'll improve the explanation in a longer note